### PR TITLE
Update surface fitting formulation

### PR DESCRIPTION
**Pull request by kmittal2**
- Update formulation to be pointwise
- Set surface fitting term's weight adaptively
- Enable surface fitting error based termination of TMOP solver

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2825](https://github.com/mfem/mfem/pull/2825) | @kmittal2 | @tzanio | @tzanio + @vladotomov | 2/15/22 | 3/12/22 | 3/15/22 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/tmop_tools.cpp [hunk 1]
```diff
@@ -638,6 +747,47 @@ void TMOPNewtonSolver::ProcessNewState(const Vector &x) const
          }
       }
    }
+
+   if (update_surf_fit_coeff)
+   {
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:752]: You should add 2 sentences about the general idea of the adaptive weight, or add more comments to the important lines below.

File: fem/tmop.cpp [hunk 2]
```diff
@@ -3151,56 +3168,47 @@ void TMOP_Integrator::AssembleElemVecSurfFit(const FiniteElement &el_x,
    const FiniteElement &el_s = *surf_fit_gf->FESpace()->GetFE(el_id);
```
Comment (vladotomov) [inline fem/tmop.cpp:3168]: Function params `ir_quad` and `weights` can be removed, and the same for the `GradSurfFit` function.

File: fem/tmop.cpp [hunk 3]
```diff
@@ -3151,56 +3168,47 @@ void TMOP_Integrator::AssembleElemVecSurfFit(const FiniteElement &el_x,
...
    surf_fit_gf->FESpace()->GetElementDofs(el_id, dofs);
-   surf_fit_gf->GetSubVector(dofs, surf_fit_gf_e);
-   surf_fit_gf_bar->GetSubVector(dofs, surf_fit_gf_bar_e);
-   surf_fit_gf_bar->GetValues(el_id, ir_quad, surf_fit_gf_bar_q);
+   int count = 0;
+   for (int s = 0; s < dofs.Size(); s++)
+   {
+      count += ((*surf_fit_marker)[dofs[s]]) ? 1 : 0;
+   }
+   if (count == 0) { return; }
```
Comment (vladotomov) [inline fem/tmop.cpp:3181]: The `count` check can be moved earlier, and the same for the `GradSurfFit` function.

File: fem/tmop.cpp [hunk 4]
```diff
@@ -58,6 +58,29 @@ void TMOP_Combo_QualityMetric::AssembleH(const DenseMatrix &Jpt,
    }
 }
 
+double TMOP_Metric_000::EvalW(const DenseMatrix &Jpt) const
+{
+   return 0.0;
+}
+
+void TMOP_Metric_000::EvalP(const DenseMatrix &Jpt, DenseMatrix &P) const
```
Comment (vladotomov) [inline fem/tmop.cpp:66]: I think simply `P=0` should work fine.

File: fem/tmop.cpp [hunk 5]
```diff
@@ -58,6 +58,29 @@ void TMOP_Combo_QualityMetric::AssembleH(const DenseMatrix &Jpt,
...
+{
+   ie.SetJacobian(Jpt.GetData());
+   P = ie.Get_dI1();
+   P *= 0.0;
+}
+
+void TMOP_Metric_000::AssembleH(const DenseMatrix &Jpt,
+                                const DenseMatrix &DS,
+                                const double weight,
+                                DenseMatrix &A) const
```
Comment (vladotomov) [inline fem/tmop.cpp:76]: `A=0`

File: miniapps/meshing/pmesh-optimizer.cpp [hunk 6]
```diff
@@ -414,6 +423,7 @@ int main (int argc, char *argv[])
    switch (metric_id)
    {
       // T-metrics
+      case 0: metric = new TMOP_Metric_000; break;
```
Comment (vladotomov) [inline miniapps/meshing/pmesh-optimizer.cpp:426]: Can users actually use this one?

File: miniapps/meshing/pmesh-optimizer.cpp [hunk 7]
```diff
@@ -72,7 +72,8 @@
 //  Adaptive surface fitting:
```
Comment (vladotomov) [inline miniapps/meshing/pmesh-optimizer.cpp:72]: Add corresponding changes in `mesh-optimizer`.

**Review by vladotomov — CHANGES_REQUESTED** [review]

**Review by vladotomov — APPROVED** [review]
Works great, good job @kmittal2.

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**